cpp/src/glm/qn/glm_logistic.cuh

```diff
@@ -32,7 +32,9 @@ struct LogisticLoss : GLMBase<T, LogisticLoss<T>> {
     : Super(handle, D, 1, has_bias) {}
 
   inline __device__ T log_sigmoid(T x) const {
-    return -MLCommon::myLog(1 + MLCommon::myExp(-x));
+    // To avoid floating point overflow in the exp function
+    return x < 0 ? x - MLCommon::myLog(1 + MLCommon::myExp(x))
```

Comment:
If compiler doesn't end up doing CSE, the `myLog` and `myExp` might get evaluated twice, once for threads in the "true" branch and another for the "false" branch. How about manually/explicitly grouping these common calls together to avoid this from happening?

Reply:
I do not understand your comment. There is no common subexpression: depending on the value of x we evaluate exp(-x) or exp(x). This is done to avoid fp overflow when abs(x) is large.

Reply:
In other words, the divergence is the price we have to pay to ensure numerical stability.

Reply:
maybe CSE was not the right word to be used here. What I meant to say was to rewrite this conditional as:
```
auto temp = MLCommon::myLog(1 + MLCommon::myExp(x < 0 ? x : -x));
return x < 0 ? x - temp : -temp;
```
That way, there's a high chance for compiler to issue predicated instructions as well as minimize control divergence across threads.

Reply:
Thanks for the clarification! I have modified the code accordingly.